### PR TITLE
Added grad scaler and grad clipping to updater

### PR DESCRIPTION
**Pull request by lrilling**
The following changes were made:
- Moved back propagation and optimizer step to updater
- Added grad clipping using `gradient_clip` and `gradient_clip_norm` values from config
- Moved gradient scaler for AMP to `returnn/torch/updater.py`
- Using gradient scaler in `updater.step()` to prevent clipping of unscaled gradients and interference with learning rate in optimizer (see [Pytorch - CUDA Automatic Mixed Precision examples](https://pytorch.org/docs/stable/notes/amp_examples.html#amp-examples)

### CODE REVIEW

File: returnn/torch/updater.py [hunk 1]
```diff
@@ -146,6 +158,31 @@ def get_optimizer(self):
...
 
+    def step(self, loss):
+        """
+        Executes backpropagation of the loss, handles AMP if necessary and calls the updater step.
+
+        :param torch.Tensor loss: The total loss of the current step
+        """
+        self.optimizer.zero_grad()
+
+        if self._grad_scaler is not None:
```
Comment (JackTemaki) [inline returnn/torch/updater.py:169]: You are missing the backward call if you have no scaler!

**Review by JackTemaki — CHANGES_REQUESTED** [review]

**Review by JackTemaki — APPROVED** [review]
